[REGRESSION]: test.fail() doesn't work on expectations in listeners

**Context:**
- GOOD Playwright Version: @playwright/test@1.16.3
- BAD Playwright Version: @playwright/test@1.17.0+
- Operating System: Linux

**Code Snippet**

```javascript
import { test, expect } from '@playwright/test';

test('good in all versions', async ({ page }) => {
  test.fail();
  await page.evaluate(() => console.log(''))
  expect(1).toBe(2);
});

test('bad in 1.17 and 1.18', async ({ page }) => {
  test.fail();
  page.on('console', () => {
    expect(1).toBe(2);
  });
  await page.evaluate(() => console.log(''))
});
```

**Describe the bug**

See test results in different @playwright/test versions https://github.com/lukaw3d/playwright-fail-reproduce/actions/runs/1753794024. When running `npx playwright test` with @playwright/test versions <=1.16.3, the result was `2 passed`. After 1.17.0, second test is causing the test suit to fail with
```
Error: expect(received).toBe(expected) // Object.is equality

Expected: 2
Received: 1

    at Page.<anonymous> (/home/runner/work/playwright-fail-reproduce/playwright-fail-reproduce/expect-fail.spec.ts:12:15)
    at Page.emit (node:events:390:28)
    at Proxy.<anonymous> (/home/runner/work/playwright-fail-reproduce/playwright-fail-reproduce/node_modules/playwright-core/lib/client/page.js:130:16)
    at Proxy.emit (node:events:390:28)
    at Connection.dispatch (/home/runner/work/playwright-fail-reproduce/playwright-fail-reproduce/node_modules/playwright-core/lib/client/connection.js:208:21)
    at Immediate.<anonymous> (/home/runner/work/playwright-fail-reproduce/playwright-fail-reproduce/node_modules/playwright-core/lib/inProcessFactory.js:50:83)
    at processImmediate (node:internal/timers:464:21)

  1 skipped
  1 passed (680ms)
Error: Process completed with exit code 1.
```

--- a/packages/playwright-test/src/workerRunner.ts
+++ b/packages/playwright-test/src/workerRunner.ts
@@ -94,7 +94,11 @@ export class WorkerRunner extends EventEmitter {
     // a test runner. In the latter case, the worker state could be messed up,
     // and continuing to run tests in the same worker is problematic. Therefore,
     // we turn this into a fatal error and restart the worker anyway.
-    if (this._currentTest && this._currentTest._test._type === 'test' && this._currentTest.expectedStatus !== 'failed') {
+    // The only exception is the expect() error that we still consider ok.
+    const isExpectError = (error instanceof Error) && !!(error as any).matcherResult;
+    const isCurrentTestExpectedToFail = this._currentTest?.expectedStatus === 'failed';
+    const shouldConsiderAsTestError = isExpectError || !isCurrentTestExpectedToFail;
+    if (this._currentTest && this._currentTest._test._type === 'test' && shouldConsiderAsTestError) {
       this._currentTest._failWithError(serializeError(error), true /* isHardError */);
     } else {
       // No current test - fatal error.

--- a/tests/playwright-test/basic.spec.ts
+++ b/tests/playwright-test/basic.spec.ts
@@ -426,3 +426,19 @@ test('should not reuse worker after unhandled rejection in test.fail', async ({ 
   expect(result.output).toContain(`Error: Oh my!`);
   expect(result.output).not.toContain(`Did not teardown test scope`);
 });
+
+test('should allow unhandled expects in test.fail', async ({ runInlineTest }) => {
+  const result = await runInlineTest({
+    'a.spec.ts': `
+      const { test } = pwt;
+      test('failing1', async ({}) => {
+        test.fail();
+        Promise.resolve().then(() => expect(1).toBe(2));
+        await new Promise(f => setTimeout(f, 100));
+      });
+    `
+  });
+  expect(result.exitCode).toBe(0);
+  expect(result.passed).toBe(1);
+  expect(result.output).not.toContain(`Error: expect`);
+});

